OSC 133 shell integration is missing start of command execution

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### What happened?

Ghostty 1.3.0 cannot calculate execution time when using OMP because the [`FTCS_COMMAND_EXECUTED`](https://iterm2.com/documentation-one-page.html#:~:text=FTCS_COMMAND_EXECUTED) is not written at the time the command is started. #3842 added 3 of the 4 FTCS prompt sequences, but OSC 133;C is needed to mark the start of command execution.

I was able to [repro](https://github.com/ghostty-org/ghostty/discussions/11340) this by disabling `shell_integration` in OMP and then writing the terminal sequences myself with and without the `\033]133;C\007` mark. Seems some other terminals don't need this, but to calculate command execution time by the terminal outside of OMP, terminals like Ghostty would need this. If it merely used `\033]133;B\007` - [`FTCS_COMMAND_START`](https://iterm2.com/documentation-one-page.html#:~:text=of%20shell%20prompt.-,FTCS_COMMAND_START,-OSC%20133%20;%20B) - the execution time would be useless because the user could've been sitting at the prompt for minutes or even days. OMP should write `FTCS_COMMAND_EXECUTED` for terminals to properly measure execution time.

This probably works in bash because Ghostty is injecting FTCS sequences into bash itself in addition to what OMP is setting for prompt vars.

Replaces #7376 that I opened directly from Discussions and wasn't able to pick a template. Sorry.

<details>
<summary><h3>Theme</h3></summary>

```yaml
# yaml-language-server: $schema=https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json
version: 3
console_title_template: >-
  {{ $pwd := .PWD -}}
  {{ .Shell }} in {{ with $repo := .Segments.Git -}}
    {{ if (gt (sub (len $pwd) (len $repo.Dir)) 20) -}}
      git:{{ $repo.RepoName }}/…/{{ base $pwd }}
    {{- else -}}
      git:{{ $repo.RepoName }}{{ trimPrefix $repo.Dir $pwd}}
    {{- end }}
  {{- else -}}
    {{ .Folder }}
  {{- end }}
final_space: true
shell_integration: true

palette:
  # https://colorkit.co/palette/02944e-4f0193-f1c516-eb850d-e44403-05a3e2-f7f6f5-7c7b7b-464646-101010/
  green: "#02944e"
  purple: "#4f0193"
  darkPurple: "#940248" # not in palette
  yellow: "#f1c516"
  orange: "#eb850d"
  red: "#e44403"
  blue: "#05a3e2"
  white: "#f7f6f5"
  gray: "#7c7b7b"
  darkGray: "#464646"
  black: "#101010"

blocks:
- type: prompt
  alignment: left
  segments:
  - type: shell
    style: diamond
    foreground: p:white
    background: p:purple
    options:
      mapped_shell_names:
        powershell: PS
        pwsh: PS

  - type: python
    style: powerline
    powerline_symbol: ""
    foreground: p:white
    background: p:darkPurple
    template: "  {{ if .Error }}{{ .Error }}{{ else }}{{ if .Venv }}{{ .Venv }} {{ end }}{{ .Full }}{{ end }} "

  - type: executiontime
    style: powerline
    powerline_symbol: ""
    foreground: p:black
    background: p:orange
    options:
      style: austin

  - type: status
    style: powerline
    powerline_symbol: ""
    foreground: p:white
    background: p:blue
    background_templates:
    - "{{ if gt .Code 0 }}p:red{{ end }}"
    options:
      always_enable: true

  - type: session
    style: powerline
    powerline_symbol: ""
    foreground: p:white
    background: p:blue
    template: "{{ if .SSHSession }}  {{ .UserName }} {{ end }}"

  - type: root
    style: powerline
    powerline_symbol: ""
    foreground: p:white
    background: p:red
    template: "  "

  - type: path
    style: powerline
    powerline_symbol: ""
    foreground: p:black
    background: p:gray
    template: ' {{ path .Path .Location }} {{ repeat .StackCount "" }}'
    options:
      style: full

- type: prompt
  alignment: left
  newline: true
  segments:
  - type: git
    cache:
      duration: 30s
      strategy: folder
    style: diamond
    trailing_diamond: ""
    foreground: p:white
    foreground_templates:
    - "{{ if or (.Working.Changed) (.Staging.Changed) }}p:black{{ end }}"
    - "{{ if and (gt .Ahead 0) (gt .Behind 0) }}p:white{{ end }}"
    - "{{ if gt .Ahead 0 }}p:black{{ end }}"
    background: p:green
    background_templates:
    - "{{ if or (.Working.Changed) (.Staging.Changed) }}p:yellow{{ end }}"
    - "{{ if and (gt .Ahead 0) (gt .Behind 0) }}p:red{{ end }}"
    - "{{ if gt .Ahead 0 }}p:yellow{{ end }}"
    - "{{ if gt .Behind 0 }}p:orange{{ end }}"
    template: " {{ if .UpstreamURL }}{{ url .UpstreamIcon .UpstreamURL }}{{ end }}{{
      .HEAD }}{{if .BranchStatus }} {{ .BranchStatus }}{{ end }}{{ if .Working.Changed
      }}  {{ .Working.String }}{{ end }}{{ if .Staging.Changed }}  {{ .Staging.String
      }}{{ end }} "
    options:
      branch_icon: " "
      fetch_status: true

debug_prompt:
  background: transparent
  template: "<p:white,p:red>  DBG </><p:red,transparent></> "

secondary_prompt:
  background: transparent
  template: "<p:black,p:gray> </><p:gray,transparent></> "

# vim: set et sw=2:
```

</details>

### What OS are you seeing the problem on?

macOS

### Which shell are you using?

powershell

### Log output

```Shell
Version: 29.8.0

Shell: shell (7.5.4)

Prompt:

]133;D;0]133;A shell  ]8;;file:\~]8;;\  ]133;B

Segments:

ConsoleTitle(true)                         -   0 ms
Shell(true)                                -   0 ms
Python(false)                              -   0 ms
Executiontime(false)                       -   0 ms
Status(false)                              -   1 ms
Session(false)                             -   4 ms
Root(false)                                -   0 ms
Path(true)                                 -   1 ms
Git(false)                                 -   0 ms

Run duration: 15.284834ms

Cache path: /Users/heaths/.cache/oh-my-posh

Config path: /Users/heaths/.config/oh-my-posh/theme.omp.yml

Logs:

[DEBUG] 19:03:31.958 debug.go:39 → logging enabled, raw mode: true
[DEBUG] 19:03:31.958 store.go:]:163 → (session) key not found: prompt_count_cache
[TRACE] 19:03:31.958 store.go:](session prompt_count_cache) - 31.708µs
[TRACE] 19:03:31.958 terminal.go:setPromptCount() - 34.75µs
[TRACE] 19:03:31.958 terminal.go:GOOS() - 42ns
[DEBUG] 19:03:31.958 terminal.go:setPwd:102 → /Users/heaths
[TRACE] 19:03:31.958 terminal.go:setPwd() - 314.875µs
[TRACE] 19:03:31.958 terminal.go:Init() - 366.666µs
[DEBUG] 19:03:31.958 home.go:12 → /Users/heaths
[TRACE] 19:03:31.958 path.go:Path() - 26.583µs
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading is_wsl
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading INITVERSIONPWSH
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading DSC_CONFIGURATION
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading INITVERSIONBASH
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading environment_platform
[DEBUG] 19:03:31.958 store.go:init:95 → (device) loading DSC_SHELL
[TRACE] 19:03:31.958 store.go:init(device omp.cache) - 177.584µs
[TRACE] 19:03:31.958 init.go:SessionID() - 292ns
[TRACE] 19:03:31.958 path.go:Path() - 0s
[DEBUG] 19:03:31.959 store.go:init:95 → (session) loading toggle_cache
[DEBUG] 19:03:31.959 store.go:init:95 → (session) loading CONFIG
[DEBUG] 19:03:31.959 store.go:init:95 → (session) loading CONFIG_SOURCE
[DEBUG] 19:03:31.959 store.go:init:95 → (session) loading prompt_count_cache
[DEBUG] 19:03:31.959 store.go:init:95 → (session) loading template_cache
[TRACE] 19:03:31.959 store.go:init(session pwsh.079bf02f-a857-4f92-8529-0970f7f8a775.omp.cache) - 445.917µs
[DEBUG] 19:03:31.959 load.go:isTheme:384 → /Users/heaths/.config/oh-my-posh/theme.omp.yml is not a theme
[TRACE] 19:03:31.959 load.go:resolveConfigLocation() - 11.75µs
[DEBUG] 19:03:31.959 store.go:]:176 → (device) found entry: DSC_CONFIGURATION - [0x773533e5e900]
[TRACE] 19:03:31.959 store.go:](device DSC_CONFIGURATION) - 491.541µs
[DEBUG] 19:03:31.959 home.go:12 → /Users/heaths
[DEBUG] 19:03:31.959 resource.go:Add:42 → item already exists
[TRACE] 19:03:31.960 load.go:read() - 392µs
[DEBUG] 19:03:31.960 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.960 store.go:](session toggle_cache) - 6.875µs
[DEBUG] 19:03:31.960 store.go:]:199 → (session) setting entry: toggle_cache - map[] with duration: infinite
[TRACE] 19:03:31.960 store.go:](session toggle_cache) - 2.375µs
[DEBUG] 19:03:31.960 store.go:]:199 → (device) setting entry: DSC_CONFIGURATION - [0x773533e5e900] with duration: infinite
[TRACE] 19:03:31.960 store.go:](device DSC_CONFIGURATION) - 2.625µs
[TRACE] 19:03:31.960 load.go:Parse() - 931.708µs
[TRACE] 19:03:31.960 load.go:Load() - 933.416µs
[TRACE] 19:03:31.960 terminal.go:Shell() - 84ns
[TRACE] 19:03:31.960 terminal.go:Flags() - 42ns
[DEBUG] 19:03:31.960 store.go:]:176 → (session) found entry: template_cache - {map[Path:0x773533e5a540 Shell:0x773534008be0] map[] ~ heaths  heaths Heaths-MacBook-Pro 7.5.4 pwsh /Users/heaths darwin 29.8.0 1 1 0 0 false false}
[TRACE] 19:03:31.960 store.go:](session template_cache) - 10.416µs
[TRACE] 19:03:31.960 cache.go:restoreCache() - 12.375µs
[DEBUG] 19:03:31.960 writer.go:Init:107 → terminal program: ghostty
[DEBUG] 19:03:31.960 writer.go:Init:108 → terminal shell: shell
[DEBUG] 19:03:31.960 terminal.go:Getenv:64 → NO DATA
[TRACE] 19:03:31.960 terminal.go:Getenv(OMP_CACHE_DISABLED) - 1.583µs
[DEBUG] 19:03:31.960 store.go:]:163 → (device) key not found: accent_color
[TRACE] 19:03:31.960 store.go:](device accent_color) - 2.291µs
[ERROR] 19:03:31.967 terminal.go:RunCommand:277 → exit status 1
[DEBUG] 19:03:31.967 terminal.go:RunCommand:280 ↓
    2026-03-10 19:03:31.967 defaults[64335:2275851] 
    The domain/default pair of (kCFPreferencesAnyApplication, AppleAccentColor) does not exist
[TRACE] 19:03:31.967 terminal.go:RunCommand(defaults read -g AppleAccentColor) - 7.531167ms
[ERROR] 19:03:31.967 colors_darwin.go:GetAccentColor:14 → exit status 1
[TRACE] 19:03:31.967 colors.go:SetAccentColor() - 7.541708ms
[TRACE] 19:03:31.967 terminal.go:Shell() - 83ns
[DEBUG] 19:03:31.967 terminal.go:Getenv:64 → 7.5.4
[TRACE] 19:03:31.967 terminal.go:Getenv(POSH_SHELL_VERSION) - 2.625µs
[DEBUG] 19:03:31.967 debug.go:PrintDebug:24 → segment: Title
[DEBUG] 19:03:31.967 text.go:patchTemplate:149 ↓
    {{ $pwd := .PWD -}} {{ .Shell }} in {{ with $repo := (.Segments.MustGet "Git") -}}
      {{ if (gt (sub (len $pwd) (len $repo.Dir)) 20) -}}
        git:{{ $repo.RepoName }}/…/{{ base $pwd }}
      {{- else -}}
        git:{{ $repo.RepoName }}{{ trimPrefix $repo.Dir $pwd}}
      {{- end }}
    {{- else -}}
      {{ .Folder }}
    {{- end }}
[ERROR] 19:03:31.968 concurrent.go:MustGet:36 → key Git not found
[TRACE] 19:03:31.968 terminal.go:Flags() - 41ns
[TRACE] 19:03:31.968 terminal.go:StatusCodes() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[DEBUG] 19:03:31.968 terminal.go:Getenv:64 → 50
[TRACE] 19:03:31.968 terminal.go:Getenv(POSH_CURSOR_LINE) - 1.875µs
[DEBUG] 19:03:31.968 terminal.go:Getenv:64 → 1
[TRACE] 19:03:31.968 terminal.go:Getenv(POSH_CURSOR_COLUMN) - 1.292µs
[TRACE] 19:03:31.968 terminal.go:Flags() - 41ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 41ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 84ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 41ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 83ns
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[TRACE] 19:03:31.968 terminal.go:GOOS() - 42ns
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Status
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[TRACE] 19:03:31.968 terminal.go:GOOS() - 125ns
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Python
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 5.75µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 3.458µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[TRACE] 19:03:31.968 terminal.go:StatusCodes() - 42ns
[DEBUG] 19:03:31.968 map.go:String:79 → status_template: {{ .Code }}
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[DEBUG] 19:03:31.968 map.go:String:79 → display_mode: environment
[TRACE] 19:03:31.968 terminal.go:GOOS() - 83ns
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Shell
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 12.167µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 map.go:StringArray:220 → extensions: [*.py *.ipynb pyproject.toml venv.bak]
[DEBUG] 19:03:31.968 map.go:KeyValueMap:213 → mapped_shell_names: map[powershell:PS pwsh:PS]
[DEBUG] 19:03:31.968 map.go:StringArray:220 → folders: [.venv venv virtualenv venv-win pyenv-win]
[DEBUG] 19:03:31.968 map.go:Bool:133 → home_enabled: false
[TRACE] 19:03:31.968 terminal.go:Shell() - 84ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 42ns
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[TRACE] 19:03:31.968 terminal.go:Flags() - 41ns
[TRACE] 19:03:31.968 terminal.go:GOOS() - 166ns
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Path
[TRACE] 19:03:31.968 terminal.go:GOOS() - 0s
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Executiontime
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 2.083µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.968 terminal.go:GOOS() - 42ns
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Root
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 8.375µs
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 2.291µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 text.go:patchTemplate:149 → {{ .Code }}
[TRACE] 19:03:31.968 terminal_unix.go:Root() - 708ns
[TRACE] 19:03:31.968 terminal.go:Flags() - 0s
[DEBUG] 19:03:31.968 text.go:patchTemplate:149 →  {{ .Data.Name }} 
[DEBUG] 19:03:31.968 map.go:Bool:133 → display_cygpath: false
[TRACE] 19:03:31.968 terminal.go:GOOS() - 125ns
[TRACE] 19:03:31.968 separator.go:Separator() - 42ns
[TRACE] 19:03:31.968 terminal.go:Shell() - 42ns
[TRACE] 19:03:31.968 terminal.go:GOOS() - 0s
[DEBUG] 19:03:31.968 map.go:Bool:133 → mapped_locations_enabled: true
[DEBUG] 19:03:31.968 map.go:String:79 → windows_registry_icon: 
[DEBUG] 19:03:31.968 segment.go:Execute:173 → segment: Session
[DEBUG] 19:03:31.968 map.go:String:79 → windows_registry_icon: 
[DEBUG] 19:03:31.968 home.go:12 → /Users/heaths
[TRACE] 19:03:31.968 separator.go:Separator() - 42ns
[DEBUG] 19:03:31.968 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.968 store.go:](session toggle_cache) - 19.334µs
[DEBUG] 19:03:31.968 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.968 terminal.go:Getenv:64 → NO DATA
[TRACE] 19:03:31.969 terminal.go:Getenv(SSH_CONNECTION) - 20.209µs
[DEBUG] 19:03:31.969 terminal.go:Getenv:64 → NO DATA
[TRACE] 19:03:31.969 terminal.go:Getenv(SSH_CLIENT) - 7.291µs
[TRACE] 19:03:31.969 terminal.go:GOOS() - 0s
[TRACE] 19:03:31.969 terminal.go:GOOS() - 83ns
[DEBUG] 19:03:31.969 map.go:String:79 → home_icon: ~
[DEBUG] 19:03:31.969 map.go:KeyValueMap:208 → mapped_locations: map[]
[TRACE] 19:03:31.969 separator.go:Separator() - 84ns
[TRACE] 19:03:31.969 terminal.go:GOOS() - 42ns
[TRACE] 19:03:31.969 terminal.go:GOOS() - 0s
[TRACE] 19:03:31.969 separator.go:Separator() - 41ns
[TRACE] 19:03:31.969 terminal.go:GOOS() - 41ns
[DEBUG] 19:03:31.969 store.go:]:176 → (device) found entry: environment_platform - darwin
[TRACE] 19:03:31.969 terminal.go:GOOS() - 0s
[TRACE] 19:03:31.969 terminal.go:GOOS() - 0s
[TRACE] 19:03:31.969 store.go:](device environment_platform) - 347.25µs
[TRACE] 19:03:31.969 terminal.go:GOOS() - 0s
[TRACE] 19:03:31.969 terminal.go:GOOS() - 83ns
[TRACE] 19:03:31.969 terminal.go:Flags() - 42ns
[DEBUG] 19:03:31.969 map.go:StringArray:220 → cycle: []
[DEBUG] 19:03:31.969 map.go:String:79 → folder_separator_template: 
[DEBUG] 19:03:31.969 map.go:String:79 → folder_separator_icon: /
[DEBUG] 19:03:31.969 map.go:Bool:133 → cycle_folder_separator: false
[DEBUG] 19:03:31.969 map.go:String:79 → folder_format: %s
[DEBUG] 19:03:31.969 map.go:String:79 → edge_format: %s
[DEBUG] 19:03:31.969 map.go:String:79 → left_format: %s
[DEBUG] 19:03:31.969 map.go:String:79 → right_format: %s
[TRACE] 19:03:31.969 terminal.go:Flags() - 42ns
[TRACE] 19:03:31.969 terminal.go:GOOS() - 41ns
[TRACE] 19:03:31.969 terminal.go:StackCount() - 42ns
[DEBUG] 19:03:31.969 map.go:Bool:133 → always_enabled: false
[TRACE] 19:03:31.969 terminal.go:Flags() - 83ns
[TRACE] 19:03:31.969 terminal_unix.go:DirIsWritable(/Users/heaths) - 13.292µs
[TRACE] 19:03:31.969 terminal.go:Flags() - 42ns
[DEBUG] 19:03:31.972 terminal.go:RunCommand:280 → heaths                        Mar 10 19:03
[TRACE] 19:03:31.973 terminal.go:RunCommand(who am i) - 3.607666ms
[TRACE] 19:03:31.973 terminal.go:Flags() - 42ns
[DEBUG] 19:03:31.973 text.go:patchTemplate:149 → {{ if .Data.SSHSession }}  {{ .UserName }} {{ end }}
[DEBUG] 19:03:31.973 text.go:patchTemplate:149 →  {{ path .Data.Path .Data.Location }} {{ repeat .Data.StackCount "" }}
[TRACE] 19:03:31.973 terminal.go:Shell() - 125ns
[TRACE] 19:03:31.973 terminal.go:Flags() - 0s
[TRACE] 19:03:31.973 terminal.go:Flags() - 83ns
[DEBUG] 19:03:31.973 home.go:12 → /Users/heaths
[TRACE] 19:03:31.973 terminal.go:GOOS() - 125ns
[DEBUG] 19:03:31.973 segment.go:Execute:173 → segment: Git
[DEBUG] 19:03:31.973 store.go:]:176 → (session) found entry: toggle_cache - map[]
[TRACE] 19:03:31.973 store.go:](session toggle_cache) - 4.541µs
[DEBUG] 19:03:31.973 segment.go:isToggled:291 → no toggles found
[DEBUG] 19:03:31.973 terminal.go:ResolveSymlink:195 → /Users/heaths
[TRACE] 19:03:31.973 terminal.go:ResolveSymlink(/Users/heaths) - 55.667µs
[ERROR] 19:03:31.973 terminal.go:HasParentFilePath:464 → stat .git: no such file or directory
[TRACE] 19:03:31.973 terminal.go:HasParentFilePath(.git) - 72.167µs
[DEBUG] 19:03:31.973 store.go:]:163 → (device) key not found: segment_cache_Git_/Users/heaths
[TRACE] 19:03:31.973 store.go:](device segment_cache_Git_/Users/heaths) - 3.5µs
[DEBUG] 19:03:31.973 segment.go:restoreCache:316 → no cache found for segment: Git, key: segment_cache_Git_/Users/heaths
[DEBUG] 19:03:31.973 map.go:Bool:133 → disable_with_jj: false
[DEBUG] 19:03:31.973 terminal.go:ResolveSymlink:195 → /Users/heaths
[TRACE] 19:03:31.973 terminal.go:ResolveSymlink(/Users/heaths) - 4.208µs
[ERROR] 19:03:31.973 terminal.go:HasParentFilePath:464 → stat .git: no such file or directory
[TRACE] 19:03:31.973 terminal.go:HasParentFilePath(.git) - 9.5µs
[TRACE] 19:03:31.973 terminal.go:Flags() - 0s
[TRACE] 19:03:31.973 terminal.go:Flags() - 0s
[TRACE] 19:03:31.973 terminal.go:Shell() - 42ns
[TRACE] 19:03:31.973 path.go:Path() - 42ns
```

--- a/src/cli/upgrade/notice_test.go
+++ b/src/cli/upgrade/notice_test.go
@@ -1,16 +1,57 @@
 package upgrade
 
 import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
 	"os"
 	"testing"
 
 	"github.com/jandedobbeleer/oh-my-posh/src/build"
+	rhttp "github.com/jandedobbeleer/oh-my-posh/src/runtime/http"
 	"github.com/stretchr/testify/assert"
 )
 
+// testRoundTripper redirects all outbound requests to a local test server.
+type testRoundTripper struct {
+	target *url.URL
+}
+
+func (rt *testRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	cloned := req.Clone(req.Context())
+	cloned.URL = &url.URL{
+		Scheme: rt.target.Scheme,
+		Host:   rt.target.Host,
+		Path:   req.URL.Path,
+	}
+	return http.DefaultTransport.RoundTrip(cloned)
+}
+
 func TestCanUpgrade(t *testing.T) {
+	const fakeLatest = "99.99.99"
+
+	// Serve a fake version file locally so tests never hit GitHub.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		fmt.Fprintf(w, "v%s", fakeLatest)
+	}))
+	defer server.Close()
+
+	targetURL, _ := url.Parse(server.URL)
+
+	savedHTTPClient := rhttp.HTTPClient
+	rhttp.HTTPClient = &http.Client{Transport: &testRoundTripper{target: targetURL}}
+	defer func() { rhttp.HTTPClient = savedHTTPClient }()
+
+	savedIsConnected := rhttp.IsConnected
+	rhttp.IsConnected = func() bool { return true }
+	defer func() { rhttp.IsConnected = savedIsConnected }()
+
 	ugc := &Config{}
-	latest, _ := ugc.FetchLatest()
+	latest, err := ugc.FetchLatest()
+	if err != nil {
+		t.Fatalf("failed to fetch latest version: %v", err)
+	}
 
 	cases := []struct {
 		Case           string

--- a/src/config/config.go
+++ b/src/config/config.go
@@ -149,6 +149,11 @@ func (cfg *Config) Features(env runtime.Environment) shell.Features {
 	if cfg.ShellIntegration {
 		log.Debug("shell integration enabled")
 		feats |= shell.FTCSMarks
+		// PowerShell emits FTCS_COMMAND_EXECUTED (OSC 133;C) inside the Enter key handler,
+		// so KeyHandlers must be enabled whenever shell integration is active.
+		if env.Shell() == shell.PWSH {
+			feats |= shell.KeyHandlers
+		}
 	}
 
 	// do not enable upgrade features when async is enabled

--- a/src/config/config_test.go
+++ b/src/config/config_test.go
@@ -111,6 +111,60 @@ func TestGetPalette(t *testing.T) {
 		assert.Equal(t, tc.ExpectedPalette, got, tc.Case)
 	}
 }
+func TestFeaturesShellIntegration(t *testing.T) {
+	cases := []struct {
+		Case             string
+		Shell            string
+		ShellIntegration bool
+		ExpectedFeats    shell.Features
+	}{
+		{
+			Case:             "pwsh with shell integration enables FTCSMarks and KeyHandlers",
+			Shell:            shell.PWSH,
+			ShellIntegration: true,
+			ExpectedFeats:    shell.FTCSMarks | shell.KeyHandlers,
+		},
+		{
+			Case:             "bash with shell integration enables FTCSMarks only",
+			Shell:            shell.BASH,
+			ShellIntegration: true,
+			ExpectedFeats:    shell.FTCSMarks,
+		},
+		{
+			Case:             "zsh with shell integration enables FTCSMarks only",
+			Shell:            shell.ZSH,
+			ShellIntegration: true,
+			ExpectedFeats:    shell.FTCSMarks,
+		},
+		{
+			Case:             "pwsh without shell integration enables nothing",
+			Shell:            shell.PWSH,
+			ShellIntegration: false,
+			ExpectedFeats:    0,
+		},
+	}
+
+	for _, tc := range cases {
+		env := &mock.Environment{}
+		env.On("Shell").Return(tc.Shell)
+
+		template.Cache = &cache.Template{
+			SimpleTemplate: cache.SimpleTemplate{
+				Shell: tc.Shell,
+			},
+		}
+		template.Init(env, nil, nil)
+
+		cfg := &Config{
+			ShellIntegration: tc.ShellIntegration,
+			Upgrade:          &upgrade.Config{},
+		}
+
+		got := cfg.Features(env)
+		assert.Equal(t, tc.ExpectedFeats, got, tc.Case)
+	}
+}
+
 func TestUpgradeFeatures(t *testing.T) {
 	cases := []struct {
 		Case                  string

--- a/src/runtime/http/connection.go
+++ b/src/runtime/http/connection.go
@@ -7,9 +7,9 @@ import (
 	"time"
 )
 
-// IsConnected checks if we can connect to ohmyposh within 200ms
-// If we can connect, we are connected; otherwise, let's consider being offline
-func IsConnected() bool {
+// IsConnected checks if we can connect to ohmyposh within 200ms.
+// Exposed as a variable so it can be replaced in tests.
+var IsConnected = func() bool {
 	timeout := 200 * time.Millisecond
 	dialer := &net.Dialer{
 		Timeout: timeout,

